### ISSUE TITLE
twarc2: Stuck and not downloading tweets

### ISSUE THREAD

**Issue by santoshbs** [DocNow/twarc#464]
I tried the following command:

`twarc2 search from:vinnymma --archive tmp.jsonl
`
It takes forever and then downloads nothing at all. As you can see from https://twitter.com/vinnymma, this handle has several thousand tweets. Please note that I did not ge an issue with downloading tweets for any other handles.

Not sure if there is something unique about the above handle.

**Comment by edsu** [comment]
Ah, this is interesting thanks for reporting. Even though that user is listed as having sent thousands of tweets, they appear to be automatically deleting old ones. See how far you can scroll their timeline at https://twitter.com/vinnymma and you will see.

But the behavior of twarc2 is incorrect here because it should collect what is available and then stop. But after looking in the log I can see that it is not stopping after it reached the end of the tweets and it gets rate limited.

```
2022-05-20 11:57:02,189 INFO Creating HTTP session headers for app auth.
2021-05-20 11:57:02,189 INFO getting ('https://api.twitter.com/2/tweets/search/all',) {'params': {'expansions': 'author_id,in_reply_to_user_id,referenced_tweets.id,referenced_tweets.id.author_id,entities.mentions.username,attachments.poll_ids,attachments.media_keys,geo.place_id', 'user.fields': 'created_at,description,entities,id,location,name,pinned_tweet_id,profile_image_url,protected,public_metrics,url,username,verified,withheld', 'tweet.fields': 'attachments,author_id,context_annotations,conversation_id,created_at,entities,geo,id,in_reply_to_user_id,lang,public_metrics,text,possibly_sensitive,referenced_tweets,reply_settings,source,withheld', 'media.fields': 'duration_ms,height,media_key,preview_image_url,type,url,width,public_metrics', 'poll.fields': 'duration_minutes,end_datetime,id,options,voting_status', 'place.fields': 'contained_within,country,country_code,full_name,geo,id,name,place_type', 'max_results': 500, 'query': 'from:vinnymma', 'start_time': '2006-03-21T00:00:00+00:00'}}
2021-05-20 11:57:03,255 INFO getting ('https://api.twitter.com/2/tweets/search/all',) {'params': {'expansions': 'author_id,in_reply_to_user_id,referenced_tweets.id,referenced_tweets.id.author_id,entities.mentions.username,attachments.poll_ids,attachments.media_keys,geo.place_id', 'user.fields': 'created_at,description,entities,id,location,name,pinned_tweet_id,profile_image_url,protected,public_metrics,url,username,verified,withheld', 'tweet.fields': 'attachments,author_id,context_annotations,conversation_id,created_at,entities,geo,id,in_reply_to_user_id,lang,public_metrics,text,possibly_sensitive,referenced_tweets,reply_settings,source,withheld', 'media.fields': 'duration_ms,height,media_key,preview_image_url,type,url,width,public_metrics', 'poll.fields': 'duration_minutes,end_datetime,id,options,voting_status', 'place.fields': 'contained_within,country,country_code,full_name,geo,id,name,place_type', 'max_results': 500, 'query': 'from:vinnymma', 'start_time': '2006-03-21T00:00:00+00:00', 'next_token': '1jzu9lk96gu5npw448fag6sx7pjoawd6byukc8ovktj1'}}
2021-05-20 11:57:03,511 INFO no more results for search
2021-05-20 11:57:03,511 INFO getting ('https://api.twitter.com/2/tweets/search/all',) {'params': {'expansions': 'author_id,in_reply_to_user_id,referenced_tweets.id,referenced_tweets.id.author_id,entities.mentions.username,attachments.poll_ids,attachments.media_keys,geo.place_id', 'user.fields': 'created_at,description,entities,id,location,name,pinned_tweet_id,profile_image_url,protected,public_metrics,url,username,verified,withheld', 'tweet.fields': 'attachments,author_id,context_annotations,conversation_id,created_at,entities,geo,id,in_reply_to_user_id,lang,public_metrics,text,possibly_sensitive,referenced_tweets,reply_settings,source,withheld', 'media.fields': 'duration_ms,height,media_key,preview_image_url,type,url,width,public_metrics', 'poll.fields': 'duration_minutes,end_datetime,id,options,voting_status', 'place.fields': 'contained_within,country,country_code,full_name,geo,id,name,place_type', 'max_results': 500, 'query': 'from:vinnymma', 'start_time': '2006-03-21T00:00:00+00:00', 'next_token': '1jzu9lk96gu5npw3jj1x25mca4jy57bdl2o0w5eez9fh'}}
2021-05-20 11:57:03,578 WARNING rate limit exceeded: sleeping 498.4217405319214 secs
```

twarc should be stopping when it emits this message `2021-05-20 11:57:03,511 INFO no more results for search`.

**Comment by edsu** [comment]
@SamHames do you think we need a [Twarc2.get_paginated()](https://github.com/DocNow/twarc/blob/main/twarc/client2.py#L632-L679) to stop generating pages when no more are available, or should that be handled in the code that is calling it?

**Comment by SamHames** [comment]
Hmmm, it should already be stopping - this is either a bug from the handling of the pagination tokens, OR, it might be how the search all endpoint handles deleted tweets. If there are valid `next_token` parameters in the response we should continue to iterate.

Are there lots of responses with no tweets, but a next_token parameter? If there are I think the behaviour here is correct, it's just a quirk of the API. Trying to work around it could lead to different problems.

**Comment by SamHames** [comment]
The log message isn't correct BTW - pagination is done when there's no next_token included in the response. Because tweet deletion is likely checked after retrieving the page of search results it's possible to have an empty page of results but still need to iterate.

**Comment by SamHames** [comment]
See #467 for a fix to the log messages.

**Comment by edsu** [comment]
Ah, I see what's going on now. Since there is no 'data' in the payload the sleeping behavior isn't getting trigger and it's getting immediately rate limited.

**Comment by edsu** [comment]
@santoshbs you should find this fixed in v2.0.13 which you can install with `pip3 install --upgrade twarc`
